### PR TITLE
Tickets #40 and #41: User Login and Registration

### LINKED ISSUE #40: Authentication - User Login
## Description
User needs to be able to log in again after registration

Blocked by #41

### LINKED ISSUE #41: Authentication - Registration
## Description
Users need to be able to create an accuont

## To Discuss
Can we use google or linkedin identity management? Do we need at bare minimum an option for Email/Pass?

### PR DESCRIPTION
**Pull request by a-camarillo**
#### What is the purpose of your change?
Create basic setup for user management 

- fixes #41 
- fixes #40 

#### What areas of the application does this change impact? (changes API, folder structure, etc)
- Server 

#### What features/fixes are deferred/not handled in this PR?
- Implementation of the Google Sign-In Button needs to be on the client side and when a user chooses to sign in with their google account, the token can be verified on the server side

## Checklist
- [x] Adds unit tests
- [x] Has a work item linked to the PR

### CODE REVIEW

File: server/webserver/__init__.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    from .blueprints.login import login
+
+    app.register_blueprint(login)
+
+    # @app.route("/")
+    # def homepage():
+    #     return """
+    #             <h1>This is the Homepage</h1>
+    #             <button><a href="/login">Login Here</a></button>
+    #            """
```
Comment (jwu910) [inline server/webserver/__init__.py:26]: oh yeah we're gonna set this up as the webserver too right? we'll need to sendFile() or render(), or sendHtml() or w/e it is to serve the static assets for index as a wildcard later

Comment (a-camarillo) [reply server/webserver/__init__.py:26]: oh yah, this bit of code was just to test out the Google Auth login button but I see what you mean, I think we can specify the static folder when instantiating the app e.g. `app = Flask(__name__, static_folder='/dist/')`

Comment (jwu910) [reply server/webserver/__init__.py:26]: Nice, however if we do that, we either gotta build it and commit it, or make sure CI can build and copy the artifacts over. Or oh yeah docker too

File: server/tests/functional/test_login.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    WHEN a user logs in
+    AND a POST request is made to '/login'
+    THEN verify the user
+    AND return a JWT on successful verification
+    """
+
+    # create a
+    create_user_response = client.post(
+        "/signup",
+        json={"email": "test@email.com", "username": "testUser", "password": "test"},
```
Comment (jwu910) [inline server/tests/functional/test_login.py:13]: do we/will we have min requirements for passwords? lengths, characters, etc

File: server/webserver/models.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+# callback functions
+@jwt.user_identity_loader
+def user_identity_lookup(user):
+    return user.id
+
+
+@jwt.user_lookup_loader
+def user_lookup_callback(_jwt_header, jwt_data):
+    identity = jwt_data["sub"]
```
Comment (jwu910) [inline server/webserver/models.py:26]: whats `"sub"` in this context? for clarity, could we avoid abbreviations if this is one?

File: server/webserver/queries.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    user = User(
+        email=email, name=name, hash=hash, authentication_method=authentication_method
+    )
+    db.session.add(user)
+    db.session.commit()
+    return
+
+
+def get_user(name):
+    user = db.session.execute(db.select(User).where(name=name))
```
Comment (jwu910) [inline server/webserver/queries.py:15]: will names be unique? or should we use email or ids?

**Review by jwu910 — COMMENTED** [review]
Left a few questions, will defer to @michachan for final approval tho

File: server/webserver/models.py (hunk 3 above)
Comment (a-camarillo) [reply server/webserver/models.py:26]: This is a standard abbreviation for subject specified by [RFC 7519](https://www.rfc-editor.org/rfc/rfc7519#section-4.1.2) in regards to JSON Web Tokens

File: server/tests/functional/test_login.py (hunk 2 above)
Comment (a-camarillo) [reply server/tests/functional/test_login.py:13]: This hasn't been discussed yet, length can possibly be something handled on the client side before the form data is submitted.

File: server/webserver/queries.py (hunk 4 above)
Comment (a-camarillo) [reply server/webserver/queries.py:15]: also has not been discussed, not sure if we want to display a user's full name or first and last name or just a username. I think we can alternatively require email for login. Any uniqueness can be enforced in the model definitions.

Comment (jwu910) [reply server/webserver/queries.py:15]: Do we need to keep first and last names? i think just email and password no? Theres no social aspect for this right now, and i believe no user profiles in scope

File: server/tests/functional/test_login.py (hunk 2 above)
Comment (jwu910) [reply server/tests/functional/test_login.py:13]: i think we should consider bare minimum 8 characters, 1 symbol, 1 capital letter, 1 number?

File: server/webserver/queries.py (hunk 4 above)
Comment (a-camarillo) [reply server/webserver/queries.py:15]: This is from the figma, I will message the chat to see what the plan is and if you're right we can just modify the schema
image

Comment (jwu910) [reply server/webserver/queries.py:15]: Another note, names are not unique, so we want to use a unique identifier

Comment (a-camarillo) [reply server/webserver/queries.py:15]: Sounds good, I will look at either email or id as a unique identifier for lookup. Any thoughts on having two separate queries for `get_user_by_email` and `get_user_by_id`? My thinking is that at the login/registration phase there's no way for the user to supply their id but when the client side is requesting information about a user I think it makes sense for them to request using the id.

Comment (jwu910) [reply server/webserver/queries.py:15]: Depends on use case, i think at registratioan, we've got to validate for duplicate entities anyways, (which i think you're doing). Fetching user deatils however, w/e unique identifier we decide on whether it be id or email, should be okay--as long as that information is provided as context to the client prior to any future requests.

ie:
1. User signs in
2. server responds with authenticated user payload (email/username)
3. client stores logged in user to session/state

In this case, any requests later, if necessary (if not token), client can make requests and include user email as a param if necessary

Comment (a-camarillo) [reply server/webserver/queries.py:15]: Currently using JWT so the login flow would be:

1. User signs in (client submits user form data)
2. The user info is verified on the server
3. Server sends token to client to use in subsequent requests

In step 3 I think we should send back additional information other than just the access_token. I think we can use this step to send the client the information it needs (id or email) in order to request specific user data.

Additionally, when Google auth is configured, the server can check the user information sent from Google and if the user already exists in the database we can send the same data back as the above method.

**Review by michachan — APPROVED** [review]

**Pull request merged**